src/Entity/SmsMessageResult.php

```diff
@@ -191,28 +195,30 @@ public static function baseFieldDefinitions(EntityTypeInterface $entity_type) {
   }
 
   /**
-   * Converts a plain SMS message result into an SMS message result entity.
-   *
-   * @param \Drupal\sms\Message\SmsMessageResultInterface $sms_result
-   *   A plain SMS message result.
-   *
-   * @return \Drupal\sms\Entity\SmsMessageResultInterface
-   *   An SMS message result entity that can be saved.
+   * {@inheritdoc}
    */
-  public static function convertFromMessageResult(PlainMessageResultInterface $sms_result) {
-    if ($sms_result instanceof static) {
-      return $sms_result;
+  public function preSave(EntityStorageInterface $storage) {
+    // SMS message result cannot be saved without a parent SMS message.
+    if (!$this->getSmsMessage()) {
+      throw new \LogicException('No parent SMS message specified for SMS message result');
     }
+    parent::preSave($storage);
+  }
 
-    $new = static::create();
-    $new
-      ->setCreditsBalance($sms_result->getCreditsBalance())
-      ->setCreditsUsed($sms_result->getCreditsUsed())
-      ->setError($sms_result->getError())
-      ->setErrorMessage($sms_result->getErrorMessage())
-      ->setReports($sms_result->getReports());
-
-    return $new;
+  /**
+   * {@inheritdoc}
+   */
+  public static function postLoad(EntityStorageInterface $storage, array &$entities) {
```

Comment:
Now I remember why I did this... there will be an infinite recursive loop if you set `$sms_message->setResult($previously_saved_result_whose_parent_is_sms_message)` and then call `$sms_message->getResult()`

Because `$sms_message->getReports()` will call `$previously_saved_result...->getReports()` which would have called `$previously_saved_result...->getSmsMessage()->getReports()` which is the same thing as `$sms_message->getReports()` - therefore a non-ending recursion.

Granted, this is a very edgey case, but it just occurred to me as I was inspecting the code.

I'll look for another way.